[docs-only] search service readme (#5503)
owncloud/ocis: services/search/README.md

* `SEARCH_EXTRACTOR_TIKA_TIKA_URL=http://YOUR-TIKA.URL`

When the search service can reach Tika, it begins to read out the content on demand. Note that files must be downloaded during the process, which can lead to delays with larger documents.

Note when using the Tika container and docker-compose:

* See the [ocis_wopi](https://github.com/owncloud/ocis/tree/master/deployments/examples/ocis_wopi) example.
* Containers for the linked service are reachable at a hostname identical to the alias, or the service name if no alias was specified.

## Search Functionality

The search service consists of two main parts which are file `indexing` and file `search`.

### Indexing

Everytime a resource changes its state, a corresponding event is triggered. Based on the event, the search service processes the file and adds them to its index. There are a few more steps between accepting the file and updating the index.

### Search

A query via the search service will return results based on the index created.

### State Changes which Trigger Indexing

The following state changes in the life cycle of a file are taken into account for creating or updating the index:

#### Resource Trashed

> The service looks in its index to see if the file already exists and marks it as deleted, at that moment, the file won't appear in any search request anymore.

Reviewer: ```suggestion
The service checks its index to see if the file has been processed. If an index entry exists, the file will be marked as deleted. In consequence, the file won't appear in search requests anymore.
```

The index entry stays intact.

#### Resource Restored

> That step is the counterpart of [Resource trashed](#resource-trashed), when deleting, the file isn't really removed from the index, instead the service just marks it as deleted.

Reviewer: ```suggestion
This step is the counterpart of [Resource Trashed](#resource-trashed). When a file is deleted, is isn't removed from the index, instead the service just marks it as deleted.
```

> This step undoes that and the file can be found again.

Reviewer: ```suggestion
This mark is removed when the file has been restored, and it shows up in search results again.
```

#### Resource Moved

> This comes into place whenever a file or folder is renamed or moved, the search index then updates the resource location path or starts indexing if no index has been created so far for all items affected. See [Notes](#notes) for an example.

Reviewer: ```suggestion
This comes into play whenever a file or folder is renamed or moved. The search index then updates the resource location path or starts indexing if no index has been created so far for all items affected. See [Notes](#notes) for an example.
```

#### Folder Created

> This step is always executed when a folder is created. The search extracts all necessary information and stores it in the search index

Reviewer: ```suggestion
The creation of a folder always triggers indexing. The search service extracts all necessary information and stores it in the search index
```

#### File Created

> This step is similar to [Folder created](#folder-created) with the difference that a file can contain even more valuable information. This starts to get interesting when it comes to real date content that needs to be exploited. Content extraction is part of the search service if configured.

Reviewer: ```suggestion
This case is similar to [Folder created](#folder-created) with the difference that a file can contain far more valuable information. This gets interesting but time-consuming when data content needs to be analyzed and indexed. Content extraction is part of the search service if configured.
```

#### File Version Restored

> Since ocis is capable of storing multiple versions of the same file, the search service also needs to take care of those versions.

Reviewer: ```suggestion
Since Infinite Scale is capable of storing multiple versions of the same file, the search service also needs to take care of those versions.
```

> When that step is triggered, the service starts to extract all needed information, stores them in the index and makes them discoverable.

Reviewer: ```suggestion
When a file version is restored, the service starts to extract all needed information, creates the index and makes the file discoverable.
```

#### Resource Tag Added

> Whenever a resource gets a new tag, the service takes care of it and makes that resource discoverable by that.

Reviewer: ```suggestion
Whenever a resource gets a new tag, the service takes care of it and makes that resource discoverable by the tag.
```

#### Resource Tag Removed

> This is the counterpart of [Resource tag added](#resource-tag-added), it takes care that a tag gets unassigned from the referenced resource.

Reviewer: ```suggestion
This is the counterpart of [Resource tag added](#resource-tag-added). It takes care that a tag gets unassigned from the referenced resource.
```

#### File Uploaded - Synchronous

> This step gets triggered only if `async post processing` is disabled, if it's like that,

Reviewer: ```suggestion
This case only triggers indexing if `async post processing` is disabled.
```

> the service starts to extract all needed file information, stores them in the index and makes them discoverable.

Reviewer: ```suggestion
If so, the service starts to extract all needed file information, stores it in the index and makes it discoverable.
```

#### File Uploaded - Asynchronous

This is exactly the same as [File uploaded - synchronous](#file-uploaded---synchronous) with the only difference that it is used for asynchronous uploads.

## Manually Trigger Re-Indexing a Space

> The service contains a CLI to trigger re-indexing a space:

Reviewer: ```suggestion
The service includes a command-line interface to trigger re-indexing a space:
```

```shell
ocis search index --space $SPACE_ID --user $USER_ID
```

> Note it is important that not names but ID's are necessary and that the specified user ID needs access to the space to be indexed.

Reviewer: ```suggestion
Note that not names but IDs are necessary and that the specified user ID needs access to the space to be indexed.
```

## Notes

> The indexing process tries to be self-healing for some situations.

Reviewer: ```suggestion
The indexing process tries to be self-healing in some situations.
```

> Following example, let's assume a file tree `foo/bar/baz`.  

Reviewer: ```suggestion
In the following example, let's assume a file tree `foo/bar/baz` exists.  
```
